lib/pure/strmisc.nim

```diff
@@ -85,6 +86,78 @@ proc rpartition*(s: string, sep: string): (string, string, string)
 
   return partition(s, sep, right = true)
 
+
+func parseFloatThousandSep*(str: string; sep = ','; decimalDot = '.'): float {.since: (1, 5).} =
+  ## Convenience func for `parseFloat` which allows for thousand separators,
+  ## this is designed to parse floats as found in the wild formatted for humans.
+  ##
+  ## The following assumptions and requirements must be met:
+  ## - String must not be empty.
+  ## - String must be stripped of trailing and leading whitespaces.
+  ## - `sep` must not be `'-'`.
+  ## - `decimalDot` must not be `'-'` nor `' '`.
+  ## - `sep` and `decimalDot` must be different.
+  ## - Exactly 1 separator must appear after each 3 consecutive inner digits to the left of the dot and nowhere else.
+  ##
+  ## See also:
+  ## * `parseFloat <strutils.html#parseFloat,string>`_
+  runnableExamples:
+    doAssert parseFloatThousandSep("1,000") == 1000.0
+    doAssert parseFloatThousandSep("-1,000") == -1000.0
+    doAssert parseFloatThousandSep("10,000.000") == 10000.0
+    doAssert parseFloatThousandSep("1,000,000.000") == 1000000.0
+    doAssert parseFloatThousandSep("10,000,000.000") == 10000000.0
+    doAssert parseFloatThousandSep("10.000,0", '.', ',') == 10000.0
+    doAssert parseFloatThousandSep("1'000'000,000", '\'', ',') == 1000000.0
+  assert sep != '-' and decimalDot notin {'-', ' '} and sep != decimalDot
+
+  proc raiseError(i: int; c: char; s: string) {.noinline, noreturn.} =
+    raise newException(ValueError,
+      "Invalid float containing thousand separators, invalid char '$1' at index $2 for input $3" %
+      [$c, $i, s])
+
+  if str.len > 1: # Allow "0" which is valid, equal to 0.0
+    var s = newStringOfCap(str.len)
+    var idx, successive: int
+    var afterDot, lastWasDot, lastWasSep, hasAnySep, isNegative: bool
+    for c in str:
+      if c in '0' .. '9':  # Digits
+        if hasAnySep and successive > 2:
+          raiseError(idx, c, str)
+        else:
+          s.add c
+          lastWasSep = false
+          lastWasDot = false
+          inc successive
+          inc idx
+      if c == sep:  # Thousands separator, this is NOT the dot
+        if lastWasSep or afterDot or (isNegative and idx == 1 or idx == 0):
+          raiseError(idx, c, str)
+        else:
+          lastWasSep = true # Do NOT add the Thousands separator here.
+          hasAnySep = true
+          successive = 0
+      if c == decimalDot:  # This is the dot
+        if (isNegative and idx == 1 or idx == 0) or (hasAnySep and successive != 3):  # Disallow .1
+          raiseError(idx, c, str)
+        else:
+          s.add '.' # Replace decimalDot to '.' so parseFloat can take it.
+          successive = 0
+          lastWasDot = true
+          afterDot = true
+          inc idx
```

Comment:
I don't follow the logic, `idx` can get updated a variable number of times in the loop; this will give incorrect error msgs in `raiseError(idx, c, str)` in some cases; instead why not:

`for idx, c in str: ...`

I don't know if you need idx for other purposes that would be different form that `idx`, but if so use another variable; the error msg should refer to the input (str) + position wrt input (idx) so user can easily understand errors